pkg/util/azureclient/roundtripper.go

```diff
@@ -0,0 +1,84 @@
+package azureclient
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	utillog "github.com/Azure/ARO-RP/pkg/util/log"
+)
+
+const (
+	genevaOutboundRequestsTable = "outboundRequests"
+)
+
+func NewCustomRoundTripper(next http.RoundTripper) http.RoundTripper {
+	return &customRoundTripper{
+		next: next,
+	}
+}
+
+type customRoundTripper struct {
+	next http.RoundTripper
+}
+
+func (crt *customRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	startTime := time.Now()
+
+	correlationData := api.GetCorrelationDataFromCtx(req.Context())
+	if correlationData == nil {
+		correlationData = api.CreateCorrelationDataFromReq(req)
+	}
+
+	l := updateCorrelationDataAndEnrichLog(correlationData, utillog.GetLogger(), startTime)
+	l.Info("HttpRequestStart")
+
+	res, err := crt.next.RoundTrip(req)
+
+	l = updateCorrelationDataAndEnrichLogWithResponse(correlationData, l, res, startTime)
+	l.Info("HttpRequestEnd")
+
+	return res, err
+}
+
+// updateCorrelationDataAndEnrichLog receives a non nil correlationData and updates the start time. It also returns a new logrus entry
+// updated with the new field for the LOGKIND field, corresponding to a specific custom DGrep table for outgoing requests.
+func updateCorrelationDataAndEnrichLog(correlationData *api.CorrelationData, l *logrus.Entry, startTime time.Time) *logrus.Entry {
+	correlationData.RequestTime = startTime
+
+	l = utillog.EnrichWithCorrelationData(l, correlationData)
+	l = l.WithFields(logrus.Fields{
+		"LOGKIND": genevaOutboundRequestsTable,
+	})
+
+	return l
+}
+
+// updateCorrelationDataAndEnrichLogWithResponse updates the correlation data with the values of the response and the corresponding fields
+// of the logrus entry.
+func updateCorrelationDataAndEnrichLogWithResponse(correlationData *api.CorrelationData, l *logrus.Entry, res *http.Response, startTime time.Time) *logrus.Entry {
+	correlationData.CorrelationID = ""
+	responseStatusCode := "0"
+	contentLength := "-1"
+
+	if res != nil {
+		correlationData.CorrelationID = res.Header.Get("X-Ms-Correlation-Request-Id")
+		responseStatusCode = fmt.Sprint(res.StatusCode)
+		contentLength = fmt.Sprint(res.ContentLength)
+	}
+
+	l = utillog.EnrichWithCorrelationData(l, correlationData)
+
+	fields := logrus.Fields{
+		"response_status_code":   responseStatusCode,
+		"contentLength":          contentLength,
+		"durationInMilliseconds": time.Since(startTime).Milliseconds(),
+	}
+	return l.WithFields(fields)
```

Comment:
If I understand this correctly we lose the `CorrelationID` from the logs if the response is `nil`... that seems odd to me. Why don't we want to be apply to correlate such a case?

Reply:
I was testing it (changed that a couple of times) to see the behavior. Updated that to just modify the correlation data when res is not nil. Thanks!